### PR TITLE
Fix ZIndex rendering.

### LINKED ISSUE #2714: Setting ZIndex requires window resize to work
Setting `ZIndex` via bindings or code behind doesn’t work. I made a small repro here: https://github.com/MarchingCube/Avalonia/tree/feature/zindex-repro. I also tried to invalidate parent visual, but it does not work either. Only way that actually works is resizing the window as it redraws everything.

Repro is in RenderDemo on Drawing page. I added two sliders bound to `ZIndex` of both ellipses. Could probably make an unit test out of it, but I am out of time this week.

cc: @kekekeks and @grokys as you both worked on `DeferredRenderer`

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

As #2714 describes, changes to a control's `ZIndex` or reordering the child controls of a panel did not result in a re-render. This PR addresses that.

## What is the current behavior?

When a control's `ZIndex` changes or a control is moved in a `Panel`'s children collection, the change will not be reflected until some other change, such as resizing the window.

## How was the solution implemented (if it's not obvious)?

Listen for changes to `ZIndex` and `LogicalChildren` and call a new method on the renderer: `RecalculateChildren`. In the deferred renderer, calling this method results in the child nodes of the `Panel`'s visual nodes being sorted according to the new order of controls in the panel's children, and their `ZIndex`.

## Checklist

- [x] Added unit tests (if possible)?

## Fixed issues

Fixes #2714

### CODE REVIEW

**Review by kekekeks — APPROVED** [review]
lgtm

File: src/Avalonia.Visuals/Rendering/SceneGraph/VisualNode.cs [hunk 1]
```diff
@@ -172,6 +172,37 @@ public void ReplaceDrawOperation(int index, IRef<IDrawOperation> operation)
...
+            var keys = new List<long>();
+
+            for (var i = 0; i < Visual.VisualChildren.Count; ++i)
+            {
+                var child = Visual.VisualChildren[i];
+                var zIndex = child.ZIndex;
+                keys.Add(((long)zIndex << 32) + i);
+            }
+
+            keys.Sort();
```
Comment (kekekeks) [inline src/Avalonia.Visuals/Rendering/SceneGraph/VisualNode.cs]: Actually, not OK. List.Sort is not stable

**Review by kekekeks — DISMISSED** [review]

File: src/Avalonia.Visuals/Rendering/SceneGraph/VisualNode.cs (hunk 1 above)
Comment (kekekeks) [reply src/Avalonia.Visuals/Rendering/SceneGraph/VisualNode.cs]: Yep, somewhere in the time frame of merging the scene graph, GetSortedVisualList (which was using List.Sort) got replaced by a [VisualExtension.SortByZIndex](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Visuals/VisualTree/VisualExtensions.cs#L239) which does a stable sort.

Comment (grokys) [reply src/Avalonia.Visuals/Rendering/SceneGraph/VisualNode.cs]: We're sorting on the index (`i`) as well as the zindex though, which means the sort doesn't need to be stable. In fact this code was taken from WPF which does something similar _because_ `List.Sort` isn't stable: https://github.com/dotnet/wpf/blob/master/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/panel.cs#L1030-L1035

Comment (kekekeks) [reply src/Avalonia.Visuals/Rendering/SceneGraph/VisualNode.cs]: Ah, OK

**Pull request merged**